### PR TITLE
Add -safer-matching flag to disable type-based optimization of pattern-matching

### PR DESCRIPTION
**Pull request by garrigue**
The -safer-matching option disables type-based optimizations of pattern-matching compilation.
This allows to produce a match failure if a pattern-matching was wrongly assumed to be exhaustive.
Since the exhaustiveness check for GADTs has had bugs in the past, it may be useful if you need extra security with GADTs.

Example:
```ocaml
type 'a dyn = Int : int -> int dyn | Float : float -> float dyn;;
let f (Float x) = x;;
f (Obj.magic (Int 3)) ;;
Exception: Match_failure ("//toplevel//", 1, 6).
```

### CODE REVIEW

**Comment by gasche** [comment]
You disabled one use of the `Total` information in `toplevel_handler`, but there are other uses of totality information deeper down in the pattern-matching compiler:

```ocaml
let mk_failaction_neg partial ctx def =
  match partial with
  | Partial -> [...]
  | Total -> (None, Jumps.empty)
```

```ocaml
let combine_variant [...] =
  [...]
  let fail, local_jumps =
    if
      sig_complete
      ||
      match partial with
      | Total -> true
      | _ -> false
    then
      (None, Jumps.empty)
    else
      mk_failaction_neg partial ctx def
  in
```

Do these need to be modified as well?

**Comment by garrigue** [comment]
Actually, disabling that one is sufficient.
All calls from the outside of `matching.ml` go through `toplevel_handler`, which then calls `compile_fun` with `partial` set to `Partial`.
The `total` parameter you see lower is computed by the match compiler itself, without using the type information.

It might be worth adding some extra assertions, but one would first need to be sure that `Total` and `Partial` are not generated internally.

**Review by gasche — COMMENTED** [review]
After the discussion I'm convinced that the change is correct. It is also a very safe change, it's obvious that there will not be any change if the new flag is not enabled, and rather clear that setting the flag make the compiler more conservative.

(This is probably frozen during the Sequential Glaciation.)

There is room for bikeshedding on the option name. Currently it is `-safer-matching`, which is okay, but maybe it could be clearer, for example `-untyped-matching`, what do you think?

Finally: I'm not Approving yet because there is no documentation for the change, except in the Changes section. The new option should be documented both in the manpages and the manual.

**Comment by hhugo** [comment]
Does it mean that in `let () = f ()`, we would compare `f ()` to `()` and raise match_failure if unequal ?

**Comment by gasche** [comment]
(I haven't tried the branch but) I don't think so, the exhaustivity logic in your example is different, the reasoning is that all constructors in the datatype are matched.

Think of an *untyped* programming language where datatype constructors (and records) must still be declared in type declarations before being used. (They are named in groups, and pattern-matching only supports comparing constructors coming from the same group.) In absence of naming conflicts between constructors in different declarations (those are disambiguated by type-checking before pattern-matching compilation happens), you can still reason about exhaustivity when all constructors of a declaration are covered in a pattern.

I believe that the current proposal is about the situation where the type-checker lets you omit certain constructors because typing information (coming from GADTs, maybe empty variant types as well?) shows that they are not reachable.

**Comment by garrigue** [comment]
Originally the exhaustiveness information was not propagated to the match compiler, which has its own logic to determine whether a match handles all constructors. This information was added for polymorphic variants, as it allows for more efficient compilation in some cases, and later reused for GADTs, as again it allows to remove comparisons is we know through the type that only one case (among many) is possible.
So in practice using this option should change nothing for programs using neither polymorphic variants nor GADTs. It may increase code size for programs including those, so it is intended for people who prefer to be conserrvative, or are suspicious that some miscompilation occurred.

**Comment by garrigue** [comment]
@gasche ping

File: testsuite/tests/basic-more/safer_matching.ml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+   flags += "-safer-matching"
+*)
+
+type 'a dyn = Int : int -> int dyn | Float : float -> float dyn;;
+
+let f (Float x) = x;;
+
+Format.printf "%f\n%!" (f (Float 3.1415));;
+
+try Printexc.print f (Obj.magic (Int 3)) with _ -> 0.;;
```
Comment (gasche) [inline testsuite/tests/basic-more/safer_matching.ml:11]: This testcase would be much easier to understand with:
1. a typing annotation on f (`float dyn -> float`)
2. a typing annotation on the `Obj.magic` return type

(annotating `Obj.magic` is a good practice anyway to ensure that we know what type we mean.)

**Review by gasche — COMMENTED** [review]
I still believe that this option should be mentioned in the manpage and documented in the manual (more obscure options like `-match-context-row` are documented in the manual: https://v2.ocaml.org/releases/4.14/htmlman/comp.html#s:comp-options ).

It's important for users to understand the purpose of the option -- so what you had in the Changes is useful, and also the remark that it only changes for GADTs and polymorphic variants.

**Comment by garrigue** [comment]
Added to reference manual and man pages.

**Review by gasche — APPROVED** [review]

**Comment by gasche** [comment]
Thanks! Merged.

**Pull request merged**